### ISSUE TITLE
MD5 OOB crash with specific size of data

### ISSUE THREAD

**Issue by cromachina** [Chocobo1/Hash#1]
Giving a string of 829 characters as input to the MD5 runner causes a consistent OOB error during finalize, specifically on line 304 with the call to `end`:
```cpp
addDataImpl({m_buffer.begin(), m_buffer.end()});
```
This may be an off by 1 error.

Repro case:
Consider the following, where `str` is of type `std::string` and is filled with 829 characters of anything:
```cpp
Chocobo1::MD5 md5;
md5.addData(str.data(), str.length());
md5.finalize();
return md5.toString();
```

**Comment by cromachina** [comment]
Also happens on SHA2_256 in the same place.

**Comment by Chocobo1** [comment]
I confirm the problem but I cannot reproduce the crash.
Here is a possible fix: try replace all `addDataImpl({m_buffer.begin(), m_buffer.end()});` with `addDataImpl({m_buffer.data(), m_buffer.size()});`. If this fixes the crash I'll push an update.

**Comment by cromachina** [comment]
Ah sorry, Visual Studio's debugger can detect it and force a crash. I'll give this a shot though.

**Comment by cromachina** [comment]
I've verified the fix. It does not hit the OOB issue anymore for MD5.

**Comment by Chocobo1** [comment]
Thank you, fixed in 4231aa8166b24ea5246f8edc7253b5aa2aa8fec.
